src/interfaces/controllers/insert-trailer-registration/insert-trailer-registration.ts

```diff
@@ -0,0 +1,67 @@
+import { Request, Response, NextFunction } from 'express';
+import * as domain from '../../../domain';
+import { DataAccess } from '../../../utils/data-access';
+import { IRequestHandler } from '../i-request-handler';
+import { insertValidator } from './insert-trailer-validator';
+
+export class InsertTrailerRegistration implements IRequestHandler {
+  private validate(payload?: domain.TrailerRegistration): string {
+    const errors = insertValidator.validate(payload, { abortEarly: false }).error;
+    return errors?.message;
+  }
+
+  public async call(req: Request, res: Response, next: NextFunction) {
+    const trailerRegistration = req.body as domain.TrailerRegistration;
+
+    const errors = this.validate(trailerRegistration);
+    if (errors) {
+      next(new domain.HTTPError(400, errors));
+      return;
+    }
+
+    const { vin, make, certificateIssueDate } = trailerRegistration;
+    const vinOrChassisWithMake = vin.length === 17 ? vin : vin + make;
+
+    try {
+      const existingTrailerRegistration = await this.checkForExistingTrailerRegistration(vinOrChassisWithMake);
+      if (existingTrailerRegistration) {
+        const { archive, reasonForDeregistration, deregisterDate } = existingTrailerRegistration;
+        delete existingTrailerRegistration.archive;
+        trailerRegistration.archive = archive || [];
+        // TODO: Discuss with team about this scenario
+        const recordToArchive = reasonForDeregistration || deregisterDate
+          ? existingTrailerRegistration
+          : this.deregisterTrailer(existingTrailerRegistration, certificateIssueDate);
+        trailerRegistration.archive.push(recordToArchive);
+      }
+      trailerRegistration.vinOrChassisWithMake = vinOrChassisWithMake;
+      await DataAccess.getInstance().upsertTrailerRegistration(trailerRegistration);
+    } catch (error) {
+      next(new domain.HTTPError(500, error));
+      return;
+    }
+    res.status(204).send();
+  }
+
+  private async checkForExistingTrailerRegistration(
```

Comment:
it seems like we are also duplicating code for that and the only difference seems to be on the DAO (getByVinOrChassisWithMake VS getByTrn) call so maybe we can make it more generic?

Reply:
I would suggest to leave as it is only two lines of code.